### PR TITLE
Re-configure code quality tooling (eslint and prettier)

### PR DESCRIPTION
**Pull request by rkdavidson**
# Changes or Updates
Following the general setup in [this blog post](https://thomlom.dev/setup-eslint-prettier-react/), re-establishes code quality tooling so we can more easily write consistent, standardized code as a team.

**Specifically, this PR:**
- Updates the `eslint` configuration
- Adds `prettier` for code formatting (along with secondary packages to make the it play nicely with `eslint`)
- Adds `husky` and `lint-stage` to automatically fix lint/formatting issues in staged files via _pre-commit_ git hook
- Re-organizes the `package.json` a bit, and adds code tooling related scripts
- — _☝️ most of these changes are in [these two commits](https://github.com/christfellowshipchurch/web-app/pull/147/files/b959b5a6fa73ccb5b28ac3ee99f02fac09956a4c) )_ — 
- All files automatically fixed via the `npm run fix-code` in commit 3f9d31b ( 😱 large diff!)
- Creates `.vscode/extensions.json` so team members using VSCode can see recommended extensions for the project (currently, [dbaeumer.vscode-eslint](https://marketplace.visualstudio.com/items?itemName=dbaeumer.vscode-eslint) and [esbenp.prettier-vscode](https://marketplace.visualstudio.com/items?itemName=esbenp.prettier-vscode))

**General Eslint/Prettier Rules**
Team feedback very much welcome on all rules! Combined existing settings and gleaned some from `apollos-app` for consistency across web + mobile clients.

- **Max line length:** 90
- **Single quotes:** ✅
- **Semi-colons:** ✅
- **[React] Missing PropTypes**: ⚠️ (Warns)
- **[React Hooks] Exhaustive dependencies for useEffect**: ❌ (Disabled)
  - This is a smart ideal, but auto-fixed code had potential unintended functionality changes
- **Imports Ordering**: Matches `apollos-app` ("builtin", "external", "internal", etc)

# Tradeoffs
Some rules had lots of violations (mainly `react/prop-types`), so the `eslint` configuration in this PR tries to balance short and long term needs. In the case of `react/prop-types`, it is set to **warn** now but IMO long term, **error** is preferable.
This PR doesn't try to fix _all_ existing issues that require a human eye.

This PR is slightly biased towards VS Code-friendly processes, but the enforcement of rules at the `pre-commit` git hook level is intended to mitigate that bias. A developer's choice of editor/IDE shouldn't negatively impact their ability to deliver code consistent with team standards.

# Tests
- [X] ~needed snapshots updated~
- [X] ~mobile tested~
- [X] latest update from master

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -1,27 +1,36 @@
...
-        "sourceType": "module"
-    },
-    "plugins": [
-        "react"
-    ],
-    "rules": {
-    }
-}
\ No newline at end of file
+    "ecmaVersion": 2018,
```
[the remark this answers is not in the document]
Comment (rkdavidson) [reply .eslintrc.json]: Great catch, on these freebies from extending `react-app`! 😅
Fixed in 8861ecb.

File: .eslintrc.json [hunk 2]
```diff
@@ -1,27 +1,36 @@
 {
-    "env": {
-        "browser": true,
-        "es6": true,
-        "node": true
+  "env": {
+    "browser": true,
+    "es6": true,
+    "node": true
```
[the remark this answers is not in the document]
Comment (rkdavidson) [reply .eslintrc.json]: Fixed in 8861ecb

**Comment by rkdavidson** [comment]
> I'm consistently amazed at how complicated it is to setup things like babel and eslint

SAME! 😩

> I'm seeing a lot of visual code formatting, important for readability/scanning, but not a lot of rules or setup around code structure, important for consistent thought and process.

Excellent callout, and honestly this was an oversight/miss on my part. I forgot to re-add `airbnb` to the `extends` list before opening this. After trying it locally though, you preemptively called the reason I don't think we can add this just yet: over 1,400 errors that are probably trivial to resolve, but tedious and wide in scope:



That said, 100% agreed that the structural and best practices rules are a significant chunk of a linter's value.
So this is a thorn. How about I add a todo in **Fixes & Improvements** scope to add `airbnb` and, to the extent possible in one mega-monster-PR, hope to get back to 0 errors and warnings before next cycle starts?

I lean against forcing errors to be warnings "temporarily", since it's too easy to live with noisy warnings, but I defer to what you think is best for the project(s) right now!

> What is sandbox.js?

This is a custom configuration @asleepingpanda setup, somewhat similar to RN's `.ios.js` and `.android.js` build convention.
For the **web-app**, if files are named `.sandbox.js`, they are used instead _only when built for the **beta**/test environment_ (beta.christfellowship.church). I'm using this to start replacing the Events page without accidentally shipping to end users.

**Changes pushed in commit 8861ecb**
I renamed/organized some of the npm scripts, to attempt closer parity with Apollos app core.
That said, `fixlint` matches, but the others felt weird to have without a hyphen — `fixreact`, `fixprettier`, `fixcode` (which I made up to do eslint + prettier in one step 🤷‍♂️ probably unnecessary).

To make CD/CI integration simpler, I erred on side of small modular tasks:


So for example, `npm run prettier` can check for code formatting problems without touching files:

**Comment by asleepingpanda** [comment]
If it gets rid of my red lines, I love it!! 😅

**Comment by rkdavidson** [comment]
@burkeshartsis / @jordanwade There was some merge conflicts I resolved by doing `Accept all Incoming` and then re-fixing the `eslint` and `prettier` issues in those files.

**The conflicts were in:**
- src/ui/Media/MediaItem.js
- src/router/Campuses/index.js
- src/modules/CampusSelect/index.sandbox.js
- src/modules/CampusSelect/index.js
- src/metadata/index.js
- src/auth/currentUser.js

I'm confident merging this if you all are, but would feel better if you want to gut check that your recent work made it through the merge.

**Comment by burkeshartsis** [comment]
@rkdavidson I didn't make any changes to those files so I think the validation check would come from @asleepingpanda for those files.

**Comment by rkdavidson** [comment]
Was just typing up a comment about how I was going to err on the "ask for forgiveness, not permission" side for now, but realizing it's Friday afternoon and @asleepingpanda is out — which is probably the _least_ good time to risk an automatic production deploy that touches 310 files 🙃.

Will hang tight until Monday when we at least have more hands on deck for rollbacks etc should things go sideways.

But I smoke tested the deploy preview links, particularly the CampusSelect flows etc since those had conflicts, and all felt good at a glance 👍.

**Comment by rkdavidson** [comment]
Hey @asleepingpanda, thoughts on getting this merged?
Erring on the over-cautious side here, since this touches almost all user-facing frontend code.

Any last minute checks or smoke testing we should do prior to merge?

**Pull request merged**